### PR TITLE
Global tools versioning

### PR DESCRIPTION
**Pull request by tmat**
Fixes https://github.com/dotnet/arcade/issues/2234.

Allow projects to opt-in to _release only_ package versioning by setting property `PreReleaseVersionLabel` to empty string and to automatically generating assembly and file versions by setting property `AutoGenerateAssemblyVersion` to `true`.

- `AssemblyVersion` := `MAJOR`.`MINOR`.`PATCH_NUMBER_HI`.`PATCH_NUMBER_LOW`
- `FileVersion` := `AssemblyVersion`
- `PackageVersion` := `MAJOR`.`MINOR`.`PATCH_NUMBER`

where
- `R` is the number of builds the build definition started on day `YY`/`MM`/`DD`
- `SHORT_DATE` := `YY` * 1000 + `MM` * 50 + `DD`
- `PATCH_NUMBER` := (`SHORT_DATE`- `VersionBaseShortDate`) * 100 + `R` 
- `PATCH_NUMBER_HI` := `PATCH_NUMBER` / 50000
- `PATCH_NUMBER_LOW` := `PATCH_NUMBER` % 50000
- `BASE_DATE` := 19000

Note: `SHORT_DATE` is a concept used by the versioning scheme we already use.

Example for 5th build today (2019/3/14):
- `AssemblyVersion` = `FileVersion` = `3.0.0.16405`
- `PackageVersion` = `3.0.16405`

Example for 1st build on (2025/12/29):
- `AssemblyVersion` = `FileVersion` = `8.0.13.12901`
- `PackageVersion` = `8.0.662901`

### CODE REVIEW

**Comment by tmat** [comment]
@dsplaisted @nguerrera @wli3 @chcosta @JoeRobich PTAL

**Comment by tmat** [comment]
Tested on dotnet/format global tool: https://github.com/dotnet/format/pull/118
and Roslyn build task: https://github.com/dotnet/roslyn/pull/35238

**Comment by tmat** [comment]
Ping

File: src/Microsoft.DotNet.Arcade.Sdk/tools/Settings.props [hunk 1]
```diff
@@ -1,14 +1,21 @@
 <?xml version="1.0" encoding="utf-8"?>
 <!-- Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE file in the project root for more information. -->
 <Project>
+  <PropertyGroup>
+    <_ArcadeOverriddenCustomBeforeMicrosoftCommonTargets>$(CustomBeforeMicrosoftCommonTargets)</_ArcadeOverriddenCustomBeforeMicrosoftCommonTargets>
+    <_ArcadeOverriddenCustomAfterMicrosoftCommonCrossTargetingTargets>$(CustomBeforeMicrosoftCommonTargets)</_ArcadeOverriddenCustomAfterMicrosoftCommonCrossTargetingTargets>
```
Comment (DoCode) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/Settings.props]: Should that not be set to the `CustomAfterMicrosoftCommonCrossTargetingTargets` target?

Comment (tmat) [reply src/Microsoft.DotNet.Arcade.Sdk/tools/Settings.props]: Yeah, typos! Should be `_ArcadeOverriddenCustomBeforeMicrosoftCommonCrossTargetingTargets` and `CustomBeforeMicrosoftCommonCrossTargetingTargets`

**Comment by wli3** [comment]
looks good on my part

**Comment by jaredpar** [comment]
Somehow this is breaking dotnet/runtime builds now. 

```
F:\workspace\_work\1\s\.dotnet\sdk\5.0.100-preview.1.20112.7\MSBuild.dll /nologo -maxcpucount /m -verbosity:m /v:minimal /bl:F:\workspace\_work\1\s\artifacts\log\Debug\ToolsetRestore.binlog /clp:Summary /clp:ErrorsOnly;NoSummary /nr:False /p:ContinuousIntegrationBuild=True /p:TreatWarningsAsErrors=true /p:__ToolsetLocationOutputFile=F:\workspace\_work\1\s\artifacts\toolset\5.0.0-beta.20121.6.txt /t:__WriteToolsetLocation /warnaserror F:\workspace\_work\1\s\artifacts\toolset\restore.proj
F:\workspace\_work\1\s\.packages\microsoft.dotnet.arcade.sdk\5.0.0-beta.20121.6\tools\Version.BeforeCommonTargets.targets(57,5): error MSB4186: Invalid static method invocation syntax: "[MSBuild]::Multiply($(_BuildNumberYY), 1000)". Method '[MSBuild]::Multiply' not found. Static method invocation should be of the form: $([FullTypeName]::Method()), e.g. $([System.IO.Path]::Combine(`a`, `b`)). Check that all parameters are defined, are of the correct type, and are specified in the right order. [F:\workspace\_work\1\s\eng\empty.csproj]
##[error].packages\microsoft.dotnet.arcade.sdk\5.0.0-beta.20121.6\tools\Version.BeforeCommonTargets.targets(57,5): error MSB4186: Invalid static method invocation syntax: "[MSBuild]::Multiply($(_BuildNumberYY), 1000)". Method '[MSBuild]::Multiply' not found. Static method invocation should be of the form: $([FullTypeName]::Method()), e.g. $([System.IO.Path]::Combine(`a`, `b`)). Check that all parameters are defined, are of the correct type, and are specified in the right order.
(NETCORE_ENGINEERING_TELEMETRY=Category) Build failed.
##[error]BUILD: Error: Failed to generate version headers.
BUILD:     -linuxdac x64 Release -skiptests -skipbuildpackages -officialbuildid -enforcepgo
##[error]Cmd.exe exited with code '1'.
```

**Pull request merged**